ES|QL Expand support for ENRICH to the majority of what is supported by ES ingest processors

craigtaverner opened the pull request:
The ES|QL [ENRICH command](https://www.elastic.co/guide/en/elasticsearch/reference/current/esql-commands.html#esql-enrich) already supports exact match queries (based on the [`term` query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-term-query.html)) for `keyword` field matches, however as described in the documentation on how to ["enrich your data"](https://www.elastic.co/guide/en/elasticsearch/reference/current/ingest-enriching-data.html), there are three types of ENRICH supported by Elasticsearch ingest processors:

* [match](https://www.elastic.co/guide/en/elasticsearch/reference/current/match-enrich-policy-type.html) - exact keyword match based on the [`term` query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-term-query.html)
* [range](https://www.elastic.co/guide/en/elasticsearch/reference/current/range-enrich-policy-type.html) - match ranges of numbers, dates or IP addresses, based on the [`term` query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-term-query.html)
* [geo_match](https://www.elastic.co/guide/en/elasticsearch/reference/current/geo-match-enrich-policy-type.html) - geometry match based on the [`geo_shape` query](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-geo-shape-query.html)

Ideally we would like ES|QL to support all the features supported by Elasticsearch ingest. In particular there have be specific requests to support [CIDR_MATCH](https://www.elastic.co/guide/en/elasticsearch/reference/current/esql-functions-operators.html#esql-cidr_match) in ENRICH, which would be supported by the `range` enrich policy above.

Tasks:

* [x] Initial support for CIDR range match
* [ ] Optional support for `ip_range` as results column type, to improve test coverage
* [x] Support numerical range
* [x] Support date range
* [x] Support `geo_match` for intersects only.

Limitations:

* One concern would be the `geo_match` which expects the `shape_relation` to be defined in the ingest processor, not the enrich policy, which would imply that the ENRICH command needs to be enhanced to support this. However, since the `geo_match` enrich policy defaults to `intersects`, it is possible to support this without enhancing the the command syntax, and consider that enhancement as a later feature.


Fixes #106162

dnhatn committed:
Enrich skeleton

craigtaverner committed:
Initial support for CIDR range query in ENRICH

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

craigtaverner committed:
Update docs/changelog/106186.yaml

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -569,11 +569,6 @@ private LogicalPlan resolveEnrich(Enrich enrich, List<Attribute> childrenOutput)
                 if (resolved.equals(ua)) {
                     return enrich;
                 }
-                if (resolved.resolved() && resolved.dataType() != KEYWORD) {
```
We might want to bring this back for anything we don't support fully, as this provides a much nicer error message than the failure we get later.

costin replied:
Why not add it now? How many supported types are there?

craigtaverner replied:
We did add it back. My comment was from my first commit (actually Nhat's stub commit that removed it), but I brought it back almost directly, and have refined it a lot since then.

dnhatn reviewed (commented):
Some comments, but this looks great. Thanks @craigtaverner!

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/QueryList.java:
```diff
@@ -74,14 +77,33 @@ Query getQuery(int position) {
 
-    private static IntFunction<Object> blockToJavaObject(Block block) {
+    // Generate geo_match query for the input term here
+    static QueryList geoMatchQuery(MappedFieldType field, SearchExecutionContext searchExecutionContext, Block block) {
```
My intention was for this PR to include geo_match, so I've just not got round to that yet. I want IP, numerical range, date-range and geo_match. If it turns out to be too complex, then I'll split it out.

dnhatn replied:
Yes, that works for me too :)

craigtaverner committed:
Remove to_string workaround for IP range ENRICH

The original ENRICH allowed for an effective IP-range ENRICH using to_string, but this is not really the desired behavior, and was never documented nor advertised, so we'll remove it in preference to the new behaviour that directly compares IP's to ip-ranges.

craigtaverner committed:
Refined after code-review

* Moved the switch outside the lambda
* Minimized memory allocation for the byte[] used in IP decoding

craigtaverner committed:
Initial support for geo_match ENRICH

craigtaverner committed:
Fix failing tests for AnalyzerTests

craigtaverner committed:
Extended AnalyzerTests to include one CIDR ENRICH

craigtaverner committed:
Add support for integer_range

Mostly adding new tests

craigtaverner committed:
Reduced flaky test

craigtaverner committed:
Fixed failing tests

* Also removed unused file mapping-client_cidr.json since it turns out we use the file in the original location afterall
* Used boolean as the failing type, since there are no plans to ever enrich on this type, but we do want to enrich on all numbers and dates

craigtaverner committed:
Make enrich long_range tests less flaky

In CI the sharding seemed to result in variable ordering, so we changed to stats to remove that concern, as well as set fixed end-dates for ages to prevent the results changing when NOW() changes.

craigtaverner committed:
Add support for double_range in ENRICH

This already worked, so just needed tests to be updated

craigtaverner committed:
Try reduce flakiness of one test

This is likely due to the test framework issue with ENRICH indices, so we might want to remove this test, or stop doing direct tests on the source index.

craigtaverner committed:
Try reduce flakiness of one test

This is likely due to the test framework issue with ENRICH indices, so we might want to remove this test, or stop doing direct tests on the source index.

craigtaverner committed:
Try reduce flakiness of two tests

The one can get multiple results in some mixed clusters, and the other has sort-order issues in clusters.

craigtaverner committed:
Remove flaky test that only tested old behaviour

craigtaverner committed:
Expand AnalyzerTests to cover new range and geo_match queries

craigtaverner committed:
Support date_range ENRICH

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

dnhatn commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/QueryList.java:
```diff
@@ -51,54 +62,142 @@ int getPositionCount() {
+                }
+                case BYTES_REF -> {
+                    BytesRefBlock bytesRefBlock = (BytesRefBlock) block;
+                    if (field instanceof RangeFieldMapper.RangeFieldType rangeFieldType && rangeFieldType.rangeType() == RangeType.IP) {
```
I think we should support strings and ip values (IP ranges later). This means we should handle the conversion based on the data type of the input block, not the data type of the match field in the enrich index. To implement this, we'll need another prerequisite PR to include the data type of the input block in LookupRequest within EnrichLookupService.

craigtaverner replied:
I assume also we should do a validation on the data types of both the input block and the match field? Right now in Analyzer we only check that the input data type is compatible with the match type (eg. `geo_match` supports only `geo_point` and `geo_shape`), but we don't check the input type against the match field type, which we could if we passed in the match field type like you suggest.

craigtaverner replied:
I've added the input dataType support in a separate PR at https://github.com/elastic/elasticsearch/pull/106344

craigtaverner replied:
I've pushed a commit that now uses the input data type to decide on how to read the input block, which I agree makes more sense. I'm only using this for DataTypes.IP. You suggested supporting strings also. Does that mean you want to bring back support for enriching strings against IP ranges? Right now if the source index contains strings, not IP's the user would need to explicitly `EVAL ip=to_ip(str)` before doing the ENRICH. This seems like a good idea, rather that directly supporting strings inside the ENRICH.

craigtaverner replied:
As a test, I reduced the strictness on non-geo types for match/range.

This also allows keywords to be used in ip_range matches, which was the previous behaviour. However, this also means that invalid IP strings will throw errors. I added a test asserting on that error, and think this matches previous behaviour, but wonder if we should instead return null and set a warning.

dnhatn commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -587,6 +605,15 @@ private LogicalPlan resolveEnrich(Enrich enrich, List<Attribute> childrenOutput)
             return enrich;
         }
+
+        private DataType[] allowedEnrichTypes(String matchType) {
```
nit:  This can be a static map from match type to set of supported data types.

craigtaverner replied:
Agreed. Since this was not a hot path I did not both with that, but could easily refine it. I did want your opinion on the rules though, since the old code only supported keyword and ip for `match`, and I've maintained that rule, but think it might already work for the other types supported in `range`.

craigtaverner replied:
Reduced the strictness on non-geo types for match/range.

This also allows keywords to be used in ip_range matches, which was the previous behaviour. However, this also means that invalid IP strings will throw errors. I added a test asserting on that error, and think this matches previous behaviour, but wonder if we should instead return null and set a warning.

dnhatn reviewed (commented):
Looks good, thanks Craig. I have left a comment around the ip conversion. I'll do another round of review tomorrow.

dnhatn commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/QueryList.java:
```diff
@@ -51,54 +62,142 @@ int getPositionCount() {
+
+            this.field = field;
+            this.searchExecutionContext = searchExecutionContext;
+            if (block.elementType() != BYTES_REF) {
```
I think we should support ConstantNullBlock as a valid input.

craigtaverner committed:
Input null support in geo_match ENRICH

craigtaverner committed:
Use the input data type to decide how to decode the block

Previously we used the match field data type, but that is not necessarily an exact match in all cases. Since we're reading the input block, we should use the input type for decoding that block.

This only matters for blocks that need decoding. For example, ByteRefsBlock can be an encoded form of IP, Version, and all geo/spatial types.

craigtaverner committed:
Reduced the strictness on non-geo types for match/range

This also allows keywords to be used in ip_range matches, which was the previous behaviour. However, this also means that invalid IP strings will throw errors. I added a test asserting on that error, and think this matches previous behaviour, but wonder if we should instead return null and set a warning.

craigtaverner committed:
Remove IP as string check in AnalyzerTests

Since we again support matching strings to ip-range, we need to remove this test that asserts otherwise.

craigtaverner committed:
After removing the IP-string test we put instead a geo_match test

dnhatn reviewed (approved):
Thanks for adding many tests. I left some comments - most of them are optional, LGTM.

dnhatn commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -587,6 +603,13 @@ private LogicalPlan resolveEnrich(Enrich enrich, List<Attribute> childrenOutput)
         }
+
+        private static final DataType[] GEO_TYPES = new DataType[] { GEO_POINT, GEO_SHAPE };
+        private static final DataType[] NON_GEO_TYPES = new DataType[] { KEYWORD, IP, LONG, INTEGER, FLOAT, DOUBLE, DATETIME };
```
I think `TEXT` can be used for `match` and `range` types.

craigtaverner replied:
I'm thinking of refining the edge cases in followup PRs, since this one has become quite large.

dnhatn replied:
++

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvAssert.java:
```diff
@@ -239,6 +258,9 @@ public static void assertData(
                 throw ae;
             }
         }
+        if (dataFailures.isEmpty() == false) {
+            fail("Data mismatch: " + dataFailures);
+        }
```
Not sure what the goal is here however the `DataFailure` doesn't trigger the AssertError catch on line on line 234/253. And that's an issue.

craigtaverner replied:
The DataFailure approach actually provides improved details, when compared with the AssertError catch. Both approaches are trying to give more context to a failure, to make it easier to understand more about what failed, and how to fix it. In the AssertError catch code, you log the current line where the failure occurred. This will still happen for other types of failures, aside from the individual cell failure, like extra columns. The DateFailure approach tries to collect all failures together, so you can see the complete set of failures and fix them all in a bunch, instead of one at a time (which is a huge hassle for some tests). So even if this does not trigger the logging, it contains more information than that logging, so it does not matter.

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvTestsDataLoader.java:
```diff
@@ -364,11 +395,16 @@ private static void loadCsvData(
             }
         }
-        if (builder.length() > 0) {
+        if (builder.isEmpty() == false) {
```
This is a superficial change that ends up doing the same call as before but with a different comparison:
`(builder.length() == 0) == false`

craigtaverner replied:
This came from Intellij warnings. I often do 'refactoring' to reduce warnings in Intellij when working on code that has warnings. Ideally I'd like code to have no warnings.

costin commented on x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/CsvTestsDataLoader.java:
```diff
@@ -364,11 +395,16 @@ private static void loadCsvData(
     }
 
+    private static String quoteIfNecessary(String value) {
+        boolean isQuoted = (value.startsWith("\"") && value.endsWith("\"")) || (value.startsWith("{") && value.endsWith("}"));
+        return isQuoted ? value : "\"" + value + "\"";
```
could extract the double quote in a var and use it in all 4 places instead of inlining it.

costin reviewed (commented):
LGTM

craigtaverner commented on x-pack/plugin/esql/qa/testFixtures/src/main/resources/enrich-IT_tests_only.csv-spec:
```diff
@@ -151,3 +151,128 @@ row a = ["1", "2"] | enrich languages_policy on a with a_lang = language_name;
+172.21.2.162 | [Development, QA]         | 172.21.2.0/24
+;
+
+enrichAgesStatsYear#[skip:-8.13.99, reason:ENRICH extended in 8.14.0]
```
Tried this, but got the error `version [8.14.0] is not a historical version`, so this appears to only work if we're referring to older versions. That is a pity, as it was a nice improvement.

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/QueryList.java:
```diff
@@ -50,55 +62,163 @@ int getPositionCount() {
+                case LONG -> {
+                    LongBlock longBlock = (LongBlock) block;
+                    if (field instanceof RangeFieldMapper.RangeFieldType rangeFieldType && rangeFieldType.rangeType() == RangeType.DATE) {
+                        yield offset -> rangeFieldType.dateTimeFormatter().formatMillis(longBlock.getLong(offset));
```
No, the range query does something weird with the millis, setting `max=min-1` when passed a single value, which then fails the validation. Perhaps if I explicitly pass in a `max=min+1` it will bypass this?

craigtaverner committed:
Simplify DataFailures in CsvAssert

craigtaverner committed:
Use input field datatype, not enrich match datatype for DATES

craigtaverner committed:
Simplify validation of enrich types when policy is null

craigtaverner commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:
```diff
@@ -569,11 +569,6 @@ private LogicalPlan resolveEnrich(Enrich enrich, List<Attribute> childrenOutput)
                 if (resolved.equals(ua)) {
                     return enrich;
                 }
-                if (resolved.resolved() && resolved.dataType() != KEYWORD) {
```
Right now we've simply refined the Analyzer verification, but I assume you mean add something to the Verifier itself? That makes a lot of sense. Perhaps a followup PR to improve this?

craigtaverner committed:
Support doc-values for geo_point aggregations

If the enrich is before a spatial aggregation, then points can appear as doc-values.
